packages/generic-utils/src/create-deferred/create-deferred.ts

```diff
@@ -4,13 +4,20 @@
 export const createDeferred = <T>() => {
   let resolve!: (value: T | PromiseLike<T>) => void
   let reject!: (reason: any) => void
+  let settled = false
   const promise = new Promise<T>((_resolve, _reject) => {
     resolve = _resolve
     reject = _reject
   })
+
+  // we "fork" the original promise instead of directly attatching `finally` to
+  // it as it interferes with natural flow of then and catch in the user code
+  Promise.allSettled([promise]).then(() => (settled = true), void 0)
```

Comment:
Can we get rid of using allSettled here by doing something like this?
```suggestion
  Promise.all([promise]).finally(() => (settled = true), void 0)
```
Doesn't seem like there's an issue with using `finally` in general, just directly on the promise.